sdk/provisioning/Azure.Provisioning/src/ProvisioningContext.cs

```diff
@@ -0,0 +1,230 @@
+    /// accessed.
+    /// </summary>
+    /// <remarks>
+    /// Check out <see cref="DefaultAzureCredential"/> and the
+    /// <see cref="Azure.Identity"/> library for more information on getting
+    /// started with <see cref="TokenCredential"/>.
+    /// </remarks>
+    public TokenCredential DefaultArmCredential
+    {
+        get
+        {
+            TokenCredential? credential = _defaultArmCredential ?? DefaultCredential;
+            if (credential is null)
+            {
+                _defaultArmCredential = credential = DefaultCredentialProvider();
+            }
+            return credential;
+        }
+        set => _defaultArmCredential = value;
+    }
+    private TokenCredential? _defaultArmCredential;
+
+    /// <summary>
+    /// Gets or sets a <see cref="TokenCredential"/> to use for creating
+    /// <see cref="ResourceManager.ArmClient"/> instances.  If no value is
+    /// provided, it will be defaulted to a fresh value using
+    /// <see cref="DefaultCredentialProvider"/> the first time it's
+    /// accessed.
+    /// </summary>
+    /// <remarks>
+    /// Check out <see cref="DefaultAzureCredential"/> and the
+    /// <see cref="Azure.Identity"/> library for more information on getting
+    /// started with <see cref="TokenCredential"/>.
+    /// </remarks>
+    public TokenCredential DefaultClientCredential
+    {
+        get
+        {
+            TokenCredential? credential = _defaultClientCredential ?? DefaultCredential;
+            if (credential is null)
+            {
+                _defaultClientCredential = credential = DefaultCredentialProvider();
+            }
+            return credential;
+        }
+        set => _defaultClientCredential = value;
+    }
+    private TokenCredential? _defaultClientCredential;
+
+    /// <summary>
+    /// Optional callback used to configure any <see cref="ClientOptions"/>
+    /// instance used for creating <see cref="ArmClient"/> or
+    /// <see cref="ProvisioningDeployment.CreateClient"/>
+    /// instances.  This can be used to turn on logging, adjust retry timeouts,
+    /// add pipeline policies, or anything else you'd like to set up with any
+    /// Azure client.
+    /// </summary>
+    public Action<ClientOptions>? ConfigureClientOptionsCallback { get; set; }
+    #endregion Creating Clients
+
+    /// <summary>
+    /// Gets or sets a random generator.  It defaults to a new
+    /// <see cref="Random"/>.
+    /// </summary>
+    /// <remarks>
+    /// This can be helpful to set with a known seed if you want reproducible
+    /// randomness for testing or debugging.
+    /// </remarks>
+    public Random Random { get; set; } = new Random();
+
+    // TODO: Add DefaultInfrastructure class to collect resources
+    // TODO: Add a DefaultScope to allow globally scoping all resources
+    // TODO: Resource resolvers
+    // TODO: IConfig helpers
+    // TODO: Default CancellationToken to link to all async operations
```

Comment:
There are a ton of other extensibility knobs I want to dial.  Imagine an IConfig based `PropertyResolver` that looked for `Namespace:Type:Property` values and you could trivially swap your infrastructure Sku between a used Raspberry Pi 2 and an H100.

`ResourceResolver`s are another area I've been playing with that aren't quite ready yet.  The basic idea is helping people build more composable infrastructure by letting them declare what they need and either finding a suitable resource or constructing it for them.  For example, I could just say "I want to use `Blobs` and `Queues`" and have an implicit `ResourceResolver` give them a shared `StorageAccount` if it matched whatever requirements they each gave.  Ditto for every "infrastructure recipe" that just needs to stick two keys in a vault somewhere.

Finally, I really want to invest in flexible, automatic module splitting.  I think most customers are like me and not really Bicep nerds.  They're just trying desperately to get their forsaken container running and don't have the time or patience to worry about exceptions being thrown when they grab an output from one module and use it in the middle of another.  We can just handle that for people _(with extensible interfaces though so purists can still pure)_.  That opens up our `Infrastructure` class to be more of a logical grouping.  People could just start throwing them around different workload boundaries or security boundaries or however they think about their resources and we could give them helpful tools like "wrap this group of stuff in a vnet and poke one hole for this VM."  I just don't see most people doing that if they need to introduce 10 new outputs and 10 new params for every piece of data that flows across one of those new boundaries.